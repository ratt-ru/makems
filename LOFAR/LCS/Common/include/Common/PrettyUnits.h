//# PrettyUnits.h - Print units in a human-readable way
//#
//# Copyright (C) 2008
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: PrettyUnits.h 10816 2008-01-14 08:55:17Z romein $

#ifndef LOFAR_COMMON_PRETTY_UNITS_H
#define LOFAR_COMMON_PRETTY_UNITS_H

#include <string>


namespace LOFAR {

  class PrettyUnits : public std::string
  {
    protected:
      PrettyUnits(double value, const char *unit, unsigned precision);
  };

  class PrettyTime : public PrettyUnits
  {
    public:
      PrettyTime(double seconds = 0, unsigned precision = 3)
      :
	PrettyUnits(seconds, "s", precision)
      {
      }
  };

  class PrettyFrequency : public PrettyUnits
  {
    public:
      PrettyFrequency(double frequency = 0, unsigned precision = 3)
      :
	PrettyUnits(frequency, "Hz", precision)
      {
      }
  };

}  // end namespace LOFAR


#endif
