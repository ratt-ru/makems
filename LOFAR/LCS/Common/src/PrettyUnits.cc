//# PrettyUnits.cc - Print units in a human-readable way
//#
//# Copyright (C) 2008
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: PrettyUnits.cc 11104 2008-03-31 11:11:42Z romein $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <Common/PrettyUnits.h>

#include <sstream>
#include <iomanip>
#include <cmath>


namespace LOFAR {

PrettyUnits::PrettyUnits(double value, const char *unit, unsigned precision)
{
  static const char *prefixes = "yzafpnum kMGTPEZY";
  const char	    *prefix;

  if (value == 0.0)
    prefix = " ";
  else
    for (value *= 1e24, prefix = prefixes; fabs(value) >= 999.5 && prefix[1] != '\0'; prefix ++)
      value /= 1000.0;

  std::stringstream stream;
  stream << std::setprecision(precision) << std::setw(precision + 1) << value;
  *static_cast<std::string *>(this) = stream.str() + ' ' + *prefix + unit;
}

}  // end namespace LOFAR
