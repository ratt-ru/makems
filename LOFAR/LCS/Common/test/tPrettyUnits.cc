//# tTimer.cc: Test program for class NSTimer
//#
//# Copyright (C) 2004
//# ASTRON (Netherlands Foundation for Research in Astronomy)
//# P.O.Box 2, 7990 AA Dwingeloo, The Netherlands, seg@astron.nl
//#
//# This program is free software; you can redistribute it and/or modify
//# it under the terms of the GNU General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or
//# (at your option) any later version.
//#
//# This program is distributed in the hope that it will be useful,
//# but WITHOUT ANY WARRANTY; without even the implied warranty of
//# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//# GNU General Public License for more details.
//#
//# You should have received a copy of the GNU General Public License
//# along with this program; if not, write to the Free Software
//# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//#
//# $Id: tPrettyUnits.cc 11156 2008-04-16 10:18:30Z romein $

//# Always #include <lofar_config.h> first!
#include <lofar_config.h>

#include <cmath>
#include <iostream>
#include <unistd.h>
#include <Common/PrettyUnits.h>

double a = 1;

using namespace LOFAR;

int main()
{
    std::cout << PrettyTime(1) << std::endl;
    std::cout << PrettyTime(1.2345) << std::endl;
    std::cout << PrettyTime(1.23456789, 9) << std::endl;
    std::cout << PrettyTime(1.2345e-6) << std::endl;
    std::cout << PrettyTime(0) << std::endl;
    std::cout << PrettyTime(-1.2345e-6) << std::endl;
    return 0;
}
