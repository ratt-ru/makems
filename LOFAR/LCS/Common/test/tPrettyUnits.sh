#!/bin/sh
./runctest.sh tPrettyUnits 2>&1 > tPrettyUnits.log
