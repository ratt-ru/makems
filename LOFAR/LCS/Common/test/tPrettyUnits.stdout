   1  s
1.23  s
1.23456789  s
1.23 us
   0  s
-1.23 us
